### ISSUE TITLE
starlark impl of java_common.pack_sources generates invalid srcjar when inputs come from external repository

### ISSUE THREAD

**Issue by rbeasley** [bazelbuild/bazel#20882]
### Description of the bug:

Under Bazel 7, when feeding sources from an external repository to `java_common.pack_sources`, the sources are mapped to illegal paths beginning with `../`. For example, if we pass `@foo//:dummy.txt` to `pack_sources`, we'll end up with a jar containing `../foo/dummy.txt`.

### Which category does this issue belong to?

Java Rules

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```bash
## Set up main repo.
cat >pack_sources.bzl <<"EOF"
def _impl(ctx):
  java_common.pack_sources(
    ctx.actions,
    output_source_jar = ctx.outputs.out,
    sources = ctx.files.srcs,
    source_jars = [
      source_jar
      for dep in ctx.attr.deps
      for source_jar in dep[JavaInfo].source_jars
    ],
    java_toolchain = ctx.toolchains["@bazel_tools//tools/jdk:toolchain_type"].java,
  )

pack_sources = rule(
  _impl,
  attrs = {
    "deps": attr.label_list(providers=[JavaInfo]),
    "srcs": attr.label_list(allow_files = True),
  },
  outputs = {"out": "%{name}.srcjar"},
  toolchains = ["@bazel_tools//tools/jdk:toolchain_type"],
)
EOF

touch WORKSPACE

touch foo bar
cat >BUILD <<"EOF"
package(default_visibility=["//visibility:public"])
filegroup(name="srcs", srcs=["foo", "bar"])
EOF


## Set up subproject
mkdir -p subproject

cat >subproject/WORKSPACE <<"EOF"
local_repository(
  name = "superproject",
  path = "..",
)
EOF

cat >subproject/BUILD.bazel <<"EOF"
load("@superproject//:pack_sources.bzl", "pack_sources")

pack_sources(
  name = "packed",
  srcs = ["@superproject//:srcs"],
)

sh_test(
  name = "test_packed",
  srcs = ["test_packed.sh"],
  data = [":packed"],
)
EOF

cat >subproject/test_packed.sh <<"EOF"
zipinfo -1 $(find -L -name packed.srcjar) | grep -q ^foo
EOF
chmod +x subproject/test_packed.sh

## Build the srcjar and verify that foo is in the toplevel directory.
cd subproject

USE_BAZEL_VERSION=6.0.0 bazelisk test --noenable_bzlmod --java_runtime_version=remotejdk_11 :test_packed

USE_BAZEL_VERSION=7.0.0 bazelisk test --noenable_bzlmod --java_runtime_version=remotejdk_11 :test_packed
```

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 7.0.0

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

Yes. 6da7f21c427850e47dbada0e05ed690866608dd2

### Any other information, logs, or outputs that you want to share?

I suspect that the Java and Starlark implementations aren't quite the same: https://github.com/bazelbuild/bazel/blob/2f899ef8b39935d9ca131b71a24e67ef286bc378/src/main/starlark/builtins_bzl/common/java/java_helper.bzl#L295-L299 https://github.com/bazelbuild/bazel/blob/2f899ef8b39935d9ca131b71a24e67ef286bc378/src/main/java/com/google/devtools/build/lib/rules/java/SingleJarActionBuilder.java#L139-L146 

Compare `resource.getRootRelativePath()` to `file.short_path`.

IIUC, then based on the table at https://github.com/bazelbuild/rules_pkg/blob/main/pkg/mappings.bzl#L166-L177, instead of using `file.short_path` it should instead be `paths.relativize(file.path, paths.normalize(paths.join(file.root.path, file.owner.workspace_root)))`.

**Comment by fmeum** [comment]
@bazel-io fork 7.0.1

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.1 RC2](https://releases.bazel.build/7.0.1/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0
